### PR TITLE
Optimize `ownerOf` and `balanceOf`

### PR DESCRIPTION
**Pull request by Philogy**
I've made a small optimization to the `ownerOf` method by moving the `require(owner != 0)` to be within the if-block so that it's not checked when we already know it's not zero.

The optimization of the `balanceOf` method is a bit more involved. To avoid always having to call `_balanceOfPrimary` which is the expensive part of the `balanceOf` method I changed the mapping that stores the balance state to also store a flag remembering if the address already spent their primary token. This will ensure that after the first transfer of the primary token `_balanceOfPrimary` never needs to be called again.

Due to how I rearranged the `transferFrom` the cost of reverts is changed, especially the `from == owner` check which is now after most of the stateful logic meaning `WRONG_FROM` reverts are more expensive than before but I think this is a worthy tradeoff for keeping the code simpler and improving the average users' cost.

Non-primary token owners do *not* benefit from the `balanceOf` optimization yet, there's two ways how they could:
- Doing a one-time check in `transferFrom` whenever an account receives a token for the first time to check whether they're a primary owner and set flags
- A separate method that users need to call once to permanently reduce their cost.

### CODE REVIEW

**Comment by Philogy** [comment]
Note due to how the order of reverts was changed I had to add some pranks to selective tests to ensure that the expected errors are given. The modifications to the tests should not impact the meaning/utility of the tests.

**Comment by devtooligan** [comment]
This is great. I had a similar idea of storing a flag in the `_balanceOfPrimary` value, but didn't feel like making such a big change, lol.  I support going this route and closing my simpler pr.  

One thing to consider is wrapping [the final line in `balanceOf`](https://github.com/showtime-xyz/SS2ERC721/pull/7/files#diff-a1a37e427e3081b0160d102379cec87b5ad66ac7cc9762bb17904a0975bc0a3fR125) in an unchecked as its probably not possible to overflow that.

**Comment by devtooligan** [comment]
Might also be nice to add a gas snapshot in an earlier commit or else base this branch on a different branch that has a gas-snapshot to see the diff

File: src/SS2ERC721.sol [hunk 1]
```diff
@@ -284,20 +273,28 @@ abstract contract SS2ERC721 is ERC721 {
...
+        owner = _ownerOfSecondary[id];
 
-        // signed math
-        unchecked {
-            _balanceOfAdjustment[owner]--;
+        if (owner == address(0)) {
+            owner = _ownerOfPrimary(id);
+            require(owner != address(0), "NOT_MINTED");
+
+            _balanceIndicator[owner] |= NOT_PRIMARY_OWNER_FLAG;
```
Comment (0xkarmacoma) [inline src/SS2ERC721.sol:284]: I have a hard time wrapping my head around this: at this point, we know the `owner` is in fact a primary owner, right?

Comment (0xkarmacoma) [reply src/SS2ERC721.sol:284]: I think the approach is sound but the `NOT_PRIMARY_OWNER_FLAG` name is a little confusing.

_balanceIndicator[x] has the flag set when x is a primary owner that transferred their token out, and it means that the actual balance stored in the mapping is the actual running total (this is neat btw 👍)

wdyt of renaming it to something more factual like `SKIP_PRIMARY_BALANCE`?

**Comment by 0xkarmacoma** [comment]
I cherry picked your commits and merged them in this separate PR: https://github.com/showtime-xyz/SS2ERC721/pull/9

Thanks @Philogy !

**Pull request closed without merging**